x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+        final JWTClaimsSet jwtClaimsSet = signedJwt.getJWTClaimsSet();
+        final Date now = new Date();
+        final Date iat = jwtClaimsSet.getIssueTime();
+        final Date exp = jwtClaimsSet.getExpirationTime();
+        final Date nbf = jwtClaimsSet.getIssueTime();
+        final Date auth_time = jwtClaimsSet.getDateClaim("auth_time");
+        final String issuer = jwtClaimsSet.getIssuer();
+        final List<String> audiences = jwtClaimsSet.getAudience();
+        LOGGER.debug(
+            "Doing JWT validation, now=["
+                + now
+                + "], typ=["
+                + jwtHeaderType
+                + "], alg=["
+                + jwtHeaderAlgorithm
+                + "], kid=["
+                + jwsHeaderKeyID
+                + "], auth_time=["
+                + auth_time
+                + "], iat=["
+                + iat
+                + "], nbf=["
+                + nbf
+                + "], exp=["
+                + exp
+                + "], issuer=["
+                + issuer
+                + "], audiences=["
+                + audiences
+        );
+
+        // Header "typ" must be "jwt" or null
+        try {
+            JWT_HEADER_TYPE_VERIFIER.verify(jwtHeaderType, null);
+        } catch (Exception e) {
+            throw new Exception("Invalid JWT type [" + jwtHeaderType + "].", e);
+        }
+
+        // Header "alg" must be in the allowed list
+        if ((jwtHeaderAlgorithm == null) || (allowedSignatureAlgorithms.contains(jwtHeaderAlgorithm.getName()) == false)) {
+            throw new Exception(
+                "Rejected signature algorithm ["
+                    + jwtHeaderAlgorithm
+                    + "]. Allowed signature algorithms are ["
+                    + String.join(",", allowedSignatureAlgorithms)
+                    + "]"
+            );
+        }
+
+        // Probable sequence of time claims: auth_time (opt), iat (req), nbf (opt), exp (req)
+        JwtUtil.validateJwtAuthTime(allowedClockSkewSeconds, now, auth_time);
+        JwtUtil.validateJwtIssuedAtTime(allowedClockSkewSeconds, now, iat);
+        JwtUtil.validateJwtNotBeforeTime(allowedClockSkewSeconds, now, nbf);
+        JwtUtil.validateJwtExpiredTime(allowedClockSkewSeconds, now, exp);
+
+        if ((issuer == null) || (allowedIssuer.equals(issuer) == false)) {
+            throw new Exception("Rejected issuer [" + issuer + "]. Allowed issuer is [" + allowedIssuer + "]");
+        } else if ((audiences == null) || (allowedAudiences.stream().anyMatch(audiences::contains) == false)) {
+            throw new Exception("Rejected audiences [" + audiences + "]. Allowed audiences are [" + allowedAudiences + "]");
+        }
+
+        final List<JWK> jwks = jwkSet.getKeys();
+        if (jwks.isEmpty()) {
+            throw new Exception("Header did not match any JWKs.");
+        }
+
+        // Try each JWK to see if it can validate the SignedJWT signature
+        boolean verifiedSignature = false;
```

Comment:
I thought the same thing and I had it right after the algorithm check. However, @tim made a good point and asked to do issuer and audiences checks before the signature check.

Example: Configure 3 realms with different issuers. Submit a JWT from the 3rd realm. First 2 realms log failures due to signature checks. That is not really useful diagnostic information. It is more useful to show that the issuer did not match.

Same argument can be made about audiences mismatch too.

Reply:
Makes sense but still bothers me :) Can we please add a comment in the beginning of `validateSignedJwt` to say something to the effect of:

```
// We check audiences and issuer first so that we can fail early and give better diagnostic messages. None of this should be used anyhow until the JWT signature is validated below
```

Reply:
I completely refactored JWT validation into a separate class. The main method is greatly simplified for readability, and I added a comment why signature validation is at the end. Have a look at the latest commit to see what you think of the improvements.